api/agent/protocol/factory.go

```diff
@@ -27,7 +28,7 @@ func (e errorProto) Dispatch(ctx context.Context, ci CallInfo, w io.Writer) erro
 // It returns any protocol error, if present.
 type ContainerIO interface {
 	IsStreamable() bool
-
+	Version() string
```

Comment:
I don't understand why we need to expose version from the ContainerIO interface itself. it seems very much like the version will be contained within each protocol and they will be expected to write in the version when they write in a request, but nowhere in fn seems likely to do this separately here where I would guess we would: call protocol.Disptach() and if we get back some specific fdk error that the version is too high or low we cycle through all protocols by some version? the factory method to get a protocol doesn't allow specifying version? are formats themselves versioned (e.g. `format: "json1"`), in which case version here is completely gratuitous ?

since default and http are likely never to have a different version, it doesn't seem like this is anything but anticipatory for json and the construct doesn't necessarily make sense to impose on every format.

Reply:
okay, makes sense, only JSON protocol should have a version.

> if we get back some specific FDK error that the version is too high or low we cycle through all protocols by some version?

That shouldn't be the case. If function can't handle protocol version A, then Fn should do what it does - report back to a user with some verbose error.

> are formats themselves versioned (e.g. format: "json1"), in which case version here is completely gratuitous?

they are not, however, they must be, because any further changes to the protocol may break the FDK or make their request parsing inconsistent.

So, the idea of this PR to start versioning the protocols, because it is more than necessary for the FDKs to be consistent with the Fn server protocol implementation.